flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/savepoints/SavepointHandlers.java

```diff
@@ -101,6 +104,46 @@ public SavepointHandlers(@Nullable final String defaultSavepointDir) {
 		this.defaultSavepointDir = defaultSavepointDir;
 	}
 
+	/**
+	 * HTTP handler to stop a job with a savepoint.
+	 */
+	public class StopWithSavepointHandler extends TriggerHandler<RestfulGateway, StopWithSavepointRequestBody, StopWithSavepointMessageParameters> {
+
+		public StopWithSavepointHandler(
+				final GatewayRetriever<? extends RestfulGateway> leaderRetriever,
+				final Time timeout,
+				final Map<String, String> responseHeaders) {
+			super(leaderRetriever, timeout, responseHeaders, StopWithSavepointTriggerHeaders.getInstance());
+		}
+
+		@Override
+		protected CompletableFuture<String> triggerOperation(
+				final HandlerRequest<StopWithSavepointRequestBody, StopWithSavepointMessageParameters> request,
+				final RestfulGateway gateway) throws RestHandlerException {
+
+			final JobID jobId = request.getPathParameter(JobIDPathParameter.class);
+			final String requestedTargetDirectory = request.getRequestBody().getTargetDirectory();
+
+			if (requestedTargetDirectory == null && defaultSavepointDir == null) {
```

Comment:
could deduplicate this block; actually the majority of this class could be moved into a shared `AbstractSavepointTriggerHandler` class